From 40d3a328213a2112bad99410866c2ccdf9d44789 Mon Sep 17 00:00:00 2001
From: "Bernhard R. Link" <brlink@debian.org>
Date: Wed, 27 Jun 2012 18:12:44 +0200
Subject: if compiling with xft, hardcode encoding to utf-8

Unlike the X primitives, Xft does not seem to support
utf-8 when drawing a string encoded according to the current
locale. So if compiling with Xft, hardcode the encoding to be
used to utf-8.

Forwarded: yes
---
 src/events.c  |    4 ++++
 src/globals.c |   12 ++++++++----
 src/manage.c  |    4 ++++
 3 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/events.c b/src/events.c
index d973639..146adf4 100644
--- a/src/events.c
+++ b/src/events.c
@@ -774,7 +774,11 @@ selection_request (XSelectionRequestEvent *rq)
         : XStdICCTextStyle;
     }
     cl[0] = selection.text;
+#ifdef USE_XFT_FONT
+    Xutf8TextListToTextProperty(dpy, cl, 1, style, &ct);
+#else
     XmbTextListToTextProperty(dpy, cl, 1, style, &ct);
+#endif
     XChangeProperty(dpy, rq->requestor, rq->property,
                     target, 8, PropModeReplace,
                     ct.value, ct.nitems);
diff --git a/src/globals.c b/src/globals.c
index 9ebc8e3..c474584 100644
--- a/src/globals.c
+++ b/src/globals.c
@@ -285,15 +285,17 @@ rp_draw_string (rp_screen *s, Drawable d, int style, int x, int y, char *string,
                             DefaultColormap (dpy, s->screen_num));
       if (draw)
         {
-          XftDrawString8 (draw, style == STYLE_NORMAL ? &s->xft_fg_color:&s->xft_bg_color, s->xft_font, x, y, (FcChar8*) string, length);
+          XftDrawStringUtf8 (draw, style == STYLE_NORMAL ? &s->xft_fg_color:&s->xft_bg_color, s->xft_font, x, y, (FcChar8*) string, length);
           XftDrawDestroy (draw);
         }
       else
         PRINT_ERROR(("Failed to allocate XftDraw object\n"));
     }
   else
-#endif
+    Xutf8DrawString (dpy, d, defaults.font, style == STYLE_NORMAL ? s->normal_gc:s->inverse_gc, x, y, string, length);
+#else
     XmbDrawString (dpy, d, defaults.font, style == STYLE_NORMAL ? s->normal_gc:s->inverse_gc, x, y, string, length);
+#endif
 }
 
 int
@@ -310,11 +312,13 @@ rp_text_width (rp_screen *s UNUSED, XFontSet font, char *string, int count)
   if (s->xft_font)
     {
       XGlyphInfo extents;
-      XftTextExtents8 (dpy, s->xft_font, (FcChar8*) string, count, &extents);
+      XftTextExtentsUtf8 (dpy, s->xft_font, (FcChar8*) string, count, &extents);
       return extents.xOff;
     }
   else
-#endif
+    return Xutf8TextEscapement (font, string, count);
+#else
     return XmbTextEscapement (font, string, count);
+#endif
 }
 
diff --git a/src/manage.c b/src/manage.c
index 8aaec31..6bea758 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -204,7 +204,11 @@ get_wmname (Window w)
   char** cl;
 
   if (XGetWMName(dpy, w, &text_prop) != 0) {
+#ifdef USE_XFT_FONT
+    status = Xutf8TextPropertyToTextList(dpy, &text_prop, &cl, &n);
+#else
     status = XmbTextPropertyToTextList(dpy, &text_prop, &cl, &n);
+#endif
     if (status == Success && cl && n > 0) {
       name = xstrdup(cl[0]);
       XFreeStringList(cl);
